### DISCUSSION TITLE
Custom onChange with v7, how? Update migration guide?

### DISCUSSION THREAD

**Discussion by michalbundyra** [react-hook-form/react-hook-form#4977] [Q&A]
**Describe the bug**

I've migrated my one application to v7 and struggling with the other, where I have a bit more complex forms.

In the form I have 3 select boxes:
- event
- tickets
- zone

for each event we have limited number of tickets in each zone, so it's possible that some zones are full.
After selecting the event I want "ticket" select box restricted to the maximum number of available tickets and after selecting the number of tickets I'd like to see only available zones with that many spaces. (Sorry if it is confusing, but this is the business logic we want).

So let assume we have two events - with 3 and 5 max tickets.
When we select event with 5 tickets and choose 5 tickets and then we change our mind and select the other event (where only 3 tickets are available) we want to unselect number of tickets and restrict values up to 3.

The whole logic is implemented using onChange + watch and it works very nice on v6, but unfortunately it does not on v7 as we are doing there `{...register(...)}` and as per migration guide it adds onChange, onBlur, name and ref.

How we can then achieve the same result without complicating it with Controller?
Migration guide is not saying anything what we suppose to do if we have custom onChange/onBlur handlers on the inputs.

**To Reproduce**

1. Go to https://codesandbox.io/s/react-hook-form-useform-template-forked-yujy7?file=/src/index.js
2. Select Event 1
3. Select "5" tickets
4. Change to Event 2
5. Ticket select box should be cleared and max available values should be now 3. ALL GOOD
6. Change it to v7 of react-hook-form and try the same.

**Expected behavior**

We can easily provide custom onChange/onBlur listeners in v7 as it was possible in v6, alternatively migration guide + documentation should be updated, to demonstrate that case.

Thanks!

**Comment by bluebill1049** [comment]
First of all, sorry to hear the migration is difficult. If you can stay in v6, please feel confident as it's been maintained.

Back to v7, because in v7, `register` will return `onChange`, `onBlur`, `name`, and `ref`.

You will have to `register` before `render`, 

```ts
const { onChange, ...rest } = register('test')

<input {...rest} onChange={(e) => {
  onChange(e);
  handleChange(e); // your custom event 
}} />
```

**Reply by michalbundyra** [reply]
The above solution is fine, but when I have multiple fields with custom onChange/onBlur events it's getting complex. Still we can extract fields so we do not need to rename `{ onChange as onChangeEvent }` etc... but with v6 it was much simpler.

Thanks for the suggestion. I think it would be good to add above to the migration guide.

I do not have any suggestion how it can be improved for the future, v6 worked fine for me, but I guess you had reasons to change it that way.

**Reply by bluebill1049** [reply]
Yes, @michaelbirchler there is a workaround but I think it's ok to stay in v6 if it's already working unless you have specific requirements from the features for v7. V7 is not perfect so does v6, we are taking different tradeoffs.

**Reply by michalbundyra** [reply]
Thanks @bluebill1049. I just came up with the following solution:

```
<input {...register('name', {
  onChange: myCustomOnChangeEventListener,
  onBlur: myCustomOnBlurEventListener,
})} />
```

what would you say? Then there is no need to do the workaround suggested by you above. 

I can stay on v6, as it works good, but I guess v7 will be supported longer than v6, that's why I wanted to switch :)

Thanks

**Reply by bluebill1049** [reply]
V6 will be supported as long as it needs :) Just won't be adding new features.

**Reply by akhatri** [reply]
Hi both - I'm in a similar situation I've upgraded to v7 and I cant seem to work out the Select component. I have a custom `<Select/>` component that works with a controller and I couldn't find examples in the Docs. If something could be added that would be great! 

I've done a sample on Codesandbox here

https://codesandbox.io/s/lively-architecture-h3o3i?file=/src/Form.js

Basically, I want to add an onChange listener on the select when the quantity changes so I can so some server-side calculations

**Reply by bluebill1049** [reply]
`onChange={(e) => onChange(e.target.value)}` isn't this what you are missing?

**Reply by akhatri** [reply]
Hmm it didnt work for me maybe I'm not doing it correctly. I think I know the issue if I move the {...register} at the start it works, like so 
```
    <select {...register(label)} name={label} value={value} onChange={onChange}>
      {children}
    </select>
```

The updated codesandbox works so I think I have what I need! Cheers @bluebill1049 and great work on this amazing plugin!

**Reply by michalbundyra** [reply]
@bluebill1049 What do you think about the [solution I've suggested in my comment above](https://github.com/react-hook-form/react-hook-form/discussions/4977#discussioncomment-663960)?

If you happy with that solution I can find some time and try to implement it, but if you are not going to accept it, then it's no point ;-) Thanks

**Reply by bluebill1049** [reply]
hey Micahel, you want to upvote this feature request? https://github.com/react-hook-form/react-hook-form/issues/4751 I am keeping an eye on it.
